### PR TITLE
[Tizen.Applications.ComponentBased][TCSACR-398] Add ComponentPort class

### PR DESCRIPTION
**Pull request by hjhun**
### Description of Change ###
Added classes and methods will be used to communicate each other between components.
The ComponentPort API is for components of component-based-application.

### API Changes ###
 - ACR: TCSACR-398

Added:
 - string ComponentPort.PortName { get; } // Property
 - ComponentPort(string portName);
 - void ComponentPort.AddPrivilege(string privilege);
 - void ComponentPort.WaitForEvent();
 - void ComponentPort.Cancel();
 - void ComponentPort.Send(string endpoint, int timeout, object request);
 - object ComponentPort.SendSync(string endpoint, int timeout, object request);
 - abstract void ComponentPort.OnRequestEvent(string sender, object request);
 - abstract void ComponentPort.OnSyncRequestEvent(string sender, object request);
 - void ComponentPort.Dispose();

### CODE REVIEW

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        private static string LogTag = "ComponentPort";
+        private readonly string _portName;
+        private IntPtr _port = IntPtr.Zero;
+        private Interop.ComponentPort.ComponentPortRequestCallback _requestEventCallback;
+        private Interop.ComponentPort.ComponentPortSyncRequestCallback _syncRequestEventCallback;
+
+        /// <summary>
+        /// Constructor for this class.
+        /// </summary>
+        /// <exception cref="ArgumentException">Thrown when the arugment is invalid.</exception>
```
Comment (m-macias) [inline src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:43]:
```suggestion
        /// <exception cref="ArgumentException">Thrown when the argument is invalid.</exception>
```

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+            using (Parcel resParcel = new Parcel(resSafeHandle))
+            {
+                object response = FromParcel(resParcel);
+                return response;
+            }            
+        }
+
+        /// <summary>
+        /// Abstrace method for receiving a request event.
```
Comment (m-macias) [inline src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:179]:
```suggestion
        /// Abstract method for receiving a request event.
```

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        /// </summary>
+        /// <param name="sender">The name of the sender</param>
+        /// <param name="request">The serializable data</param>
+        /// <since_tizen> 9 </since_tizen>
+        protected abstract void OnRequestEvent(string sender, object request);
+
+        /// <summary>
+        /// Abstrace method for receiving a synchronous request event.
```
Comment (m-macias) [inline src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:187]:
```suggestion
        /// Abstract method for receiving a synchronous request event.
```

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+            _syncRequestEventCallback = new Interop.ComponentPort.ComponentPortSyncRequestCallback(OnSyncRequestEvent);
+            Interop.ComponentPort.SetRequestCb(_port, _requestEventCallback, IntPtr.Zero);
+            Interop.ComponentPort.SetSyncRequestCb(_port, _syncRequestEventCallback, IntPtr.Zero);
+        }
+
+        /// <summary>
+        /// Gets the port name.
+        /// </summary>
+        /// <since_tizen> 9 </since_tizen>
+        public String PortName
```
Comment (m-macias) [inline src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:65]: Each property must have exactly one 'value' section.

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        /// <since_tizen> 9 </since_tizen>
+        protected abstract void OnRequestEvent(string sender, object request);
+
+        /// <summary>
+        /// Abstrace method for receiving a synchronous request event.
+        /// </summary>
+        /// <param name="sender">The name of the sender</param>
+        /// <param name="request">The serializable data</param>
+        /// <since_tizen> 9 </since_tizen>
+        protected abstract object OnSyncRequestEvent(string sender, object request);
```
Comment (m-macias) [inline src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:192]: Method/delegate which returns non-void value must have exactly one 'returns' section.

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 1 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs]: > 


> Done

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 4 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs]: The type is wrong. I've fixed the type. It's changed to "string".
I've checked other modules, the properties don't have value section.

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 2 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs]: Done

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 3 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs]: Done

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 5 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs:193]: Done

File: src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs (hunk 4 above)
Comment (hjhun) [reply src/Tizen.Applications.ComponentBased.Port/Tizen.Applications.ComponentBased.Port/ComponentPort.cs]: If the value section is needed, please tell me again.
I'll prepare a new patchset to fix it.

**Review by pjh9216 — APPROVED** [review]

**Pull request merged**